codec/src/main/java/io/netty/handler/codec/compression/JdkZlibEncoder.java

```diff
@@ -323,7 +422,33 @@ private ChannelFuture finishEncode(final ChannelHandlerContext ctx, ChannelPromi
     private void deflate(ByteBuf out) {
         if (PlatformDependent.javaVersion() < 7) {
             deflateJdk6(out);
+        } else if (DEFLATER_DEFLATE_BYTEBUFFER != null) {
+            deflateJdk11(out);
+        } else {
+            deflateJdk7(out);
         }
+    }
+
+    private void deflateJdk11(ByteBuf out) {
+        if (out.isDirect()) {
+            try {
+                int numBytes;
+                do {
+                    numBytes = (Integer) DEFLATER_DEFLATE_BYTEBUFFER.invoke(deflater, out.nioBuffer(),
```

Comment:
you should call `nioBuffer()` only once:

```java
ByteBuffer nioBuffer = out.nioBuffer();
    int numBytes;
    do {
        numBytes = (Integer) DEFLATER_DEFLATE_BYTEBUFFER.invoke(deflater, nioBuffer, Deflater.SYNC_FLUSH);
    } while (numBytes > 0);
```

Reply:
done